src/NHibernate/Event/Default/AbstractSaveEventListener.cs

```diff
@@ -235,7 +235,7 @@ protected virtual object PerformSaveOrReplicate(object entity, EntityKey key, IE
 
 			if (persister.HasCollections)
 			{
-				substitute = substitute || VisitCollectionsBeforeSave(entity, id, values, types, source);
+				substitute = substitute | VisitCollectionsBeforeSave(entity, id, values, types, source);
```

Comment:
Better explain, to avoid a later bad optimization, even if the test case is also here to prevent the mistake to be done.

```suggestion
				// We have to visit the collections whatever the current value of substitute.
				substitute = VisitCollectionsBeforeSave(entity, id, values, types, source) || substitute;
```